File: poet/src/main/java/com/babedev/javapoet/MvpPoet.java
package com.babedev.javapoet;

import android.app.Activity;
import android.os.Bundle;

import com.squareup.javapoet.ClassName;
import com.squareup.javapoet.CodeBlock;
import com.squareup.javapoet.JavaFile;
import com.squareup.javapoet.MethodSpec;
import com.squareup.javapoet.TypeName;
import com.squareup.javapoet.TypeSpec;

import java.io.File;
import java.io.IOException;

import javax.lang.model.element.Modifier;

class MvpPoet {

    private String packageName;
    private String featureName;

    MvpPoet(String packageName, String featureName) {
        this.packageName = packageName;
        this.featureName = featureName;
    }

    void generateMvp() throws IOException {
        generateView();
        generateActivity();
        generatePresenter();
    }

    private void generateActivity() throws IOException {
        MethodSpec onCreate = MethodSpec.methodBuilder("onCreate")
                .addAnnotation(Override.class)
                .addModifiers(Modifier.PUBLIC)
                .returns(void.class)
                .addParameter(Bundle.class, "savedInstanceState")
                .addStatement("super.onCreate(savedInstanceState)")
                .addComment("TODO setContentView()")
                .addStatement("mPresenter = new $T()", getPresenterType())
                .addStatement("mPresenter.attachView(this)")
                .build();

        TypeSpec mvpActivity = TypeSpec.classBuilder(featureName + "Activity")
                .superclass(Activity.class)
                .addSuperinterface(getViewType())
                .addField(getPresenterType(), "mPresenter", Modifier.PRIVATE)
                .addMethod(onCreate)
                .build();

        JavaFile javaFile = JavaFile.builder(packageName + "." + featureName.toLowerCase(), mvpActivity)
                .build();

        File file = new File("app/src/main/java");
        javaFile.writeTo(file);
    }

    private void generateView() throws IOException {
        TypeSpec mvpView = TypeSpec.interfaceBuilder(featureName + "View")
                .build();

        JavaFile javaFile = JavaFile.builder(packageName + "." + featureName.toLowerCase(), mvpView)
                .build();

        File file = new File("app/src/main/java");
        javaFile.writeTo(file);
    }

    private void generatePresenter() throws IOException {
        MethodSpec attachView = MethodSpec.methodBuilder("attachView")
                .addJavadoc(CodeBlock.of("Attach View"))
                .addParameter(getViewType(), "view")
                .addStatement("mView = view")
                .build();

        TypeSpec mvpPresenter = TypeSpec.classBuilder(featureName + "Presenter")
                .addMethod(attachView)
                .addField(getViewType(), "mView", Modifier.PRIVATE)
                .build();

        JavaFile javaFile = JavaFile.builder(packageName + "." + featureName.toLowerCase(), mvpPresenter)
                .build();

        File file = new File("app/src/main/java");
        javaFile.writeTo(file);
    }

    private TypeName getViewType() {
        return ClassName.bestGuess(packageName + "." + featureName.toLowerCase() + "." + featureName + "View");
    }

    private TypeName getPresenterType() {
        return ClassName.bestGuess(packageName + "." + featureName.toLowerCase() + "." + featureName + "Presenter");
    }
}

File: poet/src/main/kotlin/com/babedev/kotlinpoet/MvpPoet.kt
package com.babedev.kotlinpoet

import android.app.Activity
import android.os.Bundle
import com.squareup.kotlinpoet.*

import java.io.File

class MvpPoet(private val packageName: String, private val featureName: String) {

    val rootPath = "app/src/main/kotlin"

    fun generateMvp() {
        generateView()
        generateActivity()
        generatePresenter()
    }

    private fun generateActivity() {
        val onCreate = FunSpec.builder("onCreate")
                .addModifiers(KModifier.OVERRIDE, KModifier.PUBLIC)
                .addParameter("savedInstanceState", Bundle::class)
                .addStatement("super.onCreate(savedInstanceState)")
                .addComment("TODO setContentView()")
                .addStatement("mPresenter = %T()", presenterType)
                .addStatement("mPresenter.attachView(this)")
                .build()

        val mvpActivity = TypeSpec.classBuilder(featureName + "Activity")
                .superclass(TypeName.get(Activity::class.java))
                .addSuperinterface(viewType)
                .addProperty(PropertySpec
                        .varBuilder("mPresenter", presenterType, KModifier.LATEINIT)
                        .build())
                .addFun(onCreate)
                .build()

        KotlinFile.builder(packageName + "." + featureName.toLowerCase(), featureName + "Activity")
                .addType(mvpActivity)
                .build()
                .writeTo(File(rootPath))
    }

    private fun generateView() {
        val mvpView = TypeSpec
                .interfaceBuilder(featureName + "View")
                .build()

        KotlinFile.builder(packageName + "." + featureName.toLowerCase(), featureName + "View")
                .addType(mvpView)
                .build()
                .writeTo(File(rootPath))
    }

    private fun generatePresenter() {
        val attachView = FunSpec.builder("attachView")
                .addKdoc(CodeBlock.of("Attach View"))
                .addParameter("view", viewType)
                .addStatement("mView = view")
                .build()

        val mvpPresenter = TypeSpec.classBuilder(featureName + "Presenter")
                .addFun(attachView)
                .addProperty(PropertySpec
                        .varBuilder("mView", viewType, KModifier.LATEINIT)
                        .build())
                .build()

        KotlinFile.builder(packageName + "." + featureName.toLowerCase(), featureName + "Presenter")
                .addType(mvpPresenter)
                .build()
                .writeTo(File(rootPath))
    }

    private val viewType: TypeName
        get() = ClassName.bestGuess(packageName + "." + featureName.toLowerCase() + "." + featureName + "View")

    private val presenterType: TypeName
        get() = ClassName.bestGuess(packageName + "." + featureName.toLowerCase() + "." + featureName + "Presenter")
}
